contrib/bloop/src/mill.contrib.bloop/BloopModule.scala

```diff
@@ -0,0 +1,197 @@
+package mill.contrib.bloop
+
+import ammonite.ops._
+import bloop.config.Config.{File => BloopFile}
+import bloop.config.ConfigEncoderDecoders._
+import io.circe.Encoder
+import mill._
+import mill.define._
+import mill.eval.Evaluator
+import mill.modules.Jvm
+import mill.scalalib._
+import ujson.Visitor
+import upickle.default
+
+// TODO diamond problem is going to make this tricky. Using a self-type appears
+// tobe problematic with mill. Moreover, the global "install"
+// method that inspects the build relies on whether the modules are instances of
+// JavaModule or ScalaModule, so not quite sure what approach to take to avoid
+// copy pasting between a "BloopJavaModule" and a "BloopScalaModule" (and a potential
+// BloopScalaJsModule ?)
+//
+// TODO : bloop clean doesn't seem to work for me. Also, investigate if it's
+// possible to hook with "mill clean"
+//
+trait BloopModule extends ScalaModule { self =>
+
+  trait Tests extends super.Tests with BloopModule {
+    // TODO how can we retrieve test results from bloop ?
+    override def test(args: String*) = super.test(args: _*)
+  }
+
+  override def compile = T {
+    val config = bloopConfig().project
+    val cmd = Seq("bloop", "compile", config.name)
+    val classes = Path(config.classesDir)
+    val out = Path(config.out)
+    val analysisFile = out / s"${config.name}-analysis.bin" // TODO does it need polling ?
+    Jvm.baseInteractiveSubprocess(cmd, T.ctx().env, pwd)
+    CompilationResult(analysisFile, PathRef(classes))
+  }
+
+  override def run(args: String*) = T.command {
+    val config = bloopConfig().project
+    val cmd = Seq("bloop", "run", config.name) ++ args // TODO does bloop return a specific exit code when failing to parse params ?
+    Jvm.baseInteractiveSubprocess(cmd, T.ctx().env, pwd)
+  }
+
+  def bloopConfig = T {
+    // Forcing the generation of the config for all modules here
+    val _ = BloopModule.install() // NB: looks like it has to be assigned to be executed
```

Comment:
Another TODO : this is highly inefficient. Making any change to a source file seems to invalidate the cache of "sources" and re-triggers the generation of all source files

Reply:
Good observation. I take that by change you mean change to the contents and not the timestamps. My fix to this would be to create an intermediary task that turns the `sources: List[Path]` into `sources: SourceFiles` where the hash of `SourceFiles` takes into account only the paths, but not whether they changed or not. Then we could depend on this intermediary task from Bloop's code.